src/control/lib/netdetect/netdetect.go

```diff
@@ -887,45 +1013,64 @@ func ScanFabric(provider string) ([]FabricScan, error) {
 		deviceScanCfg.depth, deviceScanCfg.numObj, deviceScanCfg.systemDeviceNames, deviceScanCfg.hwlocDeviceNames)
 
+	hfiDeviceCount := getHFIDeviceCount(deviceScanCfg.hwlocDeviceNames)
+	log.Debugf("There are %d hfi1 devices in the system", hfiDeviceCount)
+
 	for ; fi != nil; fi = fi.next {
 		if fi.domain_attr == nil || fi.domain_attr.name == nil || fi.fabric_attr == nil || fi.fabric_attr.prov_name == nil {
 			continue
 		}
-
 		deviceScanCfg.targetDevice = C.GoString(fi.domain_attr.name)
-		deviceAffinity, err := GetAffinityForDevice(deviceScanCfg)
-		if err != nil {
-			log.Debugf("Error from GetAffinityForDevice: %v", err)
-			continue
+		log.Debugf("The target device is: %s", deviceScanCfg.targetDevice)
+		// Implements a workaround to handle the current psm2 provider behavior
+		// that reports fi.domain_attr.name as "psm2" instead of an actual device
+		// name like "hfi1_0".
+		//
+		// The work around behaves in two different ways depending on the data available.
+		// When the value of fi.src_addr.unit is 0 or greater, it is used to explicitly
+		// define an hfi1_ unit by appending that number to the "hfi1_" string.
+		//
+		// When the value of fi.src_addr.unit is -1, it means that all hfi1_x units in the system
+		// are supported by the psm2 provider.  In this case, we need to know how many hfi1_x units
+		// there are, and create a fabric scan entry for each one.  By querying the device names from
+		// hwloc, we can count the number of devices matching "hfi1_" and create a device scan
+		// entry for each one.
+		if provider == "" {
+			provider = "All"
 		}
 
-		// Convert the libfabric provider list to a Mercury compatible provider list
-		mercuryProviderList, err := convertLibFabricToMercury(C.GoString(fi.fabric_attr.prov_name))
-		if err != nil {
-			log.Debugf("Couldn't convert the libfabric provider string: %s", C.GoString(fi.fabric_attr.prov_name))
-			log.Debugf("Skipping FI record for device %s, provider %s", deviceAffinity.DeviceName, C.GoString(fi.fabric_attr.prov_name))
-			// An error while converting a libfabric provider to a mercury provider is not fatal.
-			// In this case, we want to omit this libfabric record from our results because it has no
-			// mercury equivalent provider.  There are many providers in libfabric that have no mercury
-			// equivalent, and we want to filter those out right here.
-			continue
-		}
-		log.Debugf("Mercury provider list: %v", mercuryProviderList)
-
-		scanResults := FabricScan{
-			Provider:   mercuryProviderList,
-			DeviceName: deviceAffinity.DeviceName,
-			NUMANode:   deviceAffinity.NUMANode,
-			Priority:   devCount,
+		log.Debugf("This fabric info record has provider: %s", C.GoString(fi.fabric_attr.prov_name))
+		if strings.Contains(C.GoString(fi.fabric_attr.prov_name), "psm2") {
+			if strings.Contains(deviceScanCfg.targetDevice, "psm2") {
+				log.Debugf("psm2 provider and psm2 device found.")
+				hfiUnit := C.getHFIUnit(fi.src_addr)
+				switch hfiUnit {
```

Comment:
isn't this switch statement/logic a duplication and repeated above?

Reply:
There are two functions, ScanFabric() and ValidateProviderConfig() that use some elements that are similar, but have different processing logic.   There is some code that looks similar.